### PR TITLE
Week 7 Assignment

### PR DESCRIPTION
**Pull request by janewaitara**

### CODE REVIEW

**Comment by janewaitara** [comment]
Hey Jenn,
I am having this error while fetching the recipes from the api, I have tried solving it but i haven't succeeded yet.
```Rejecting re-init on previously-failed class java.lang.Class<com.janewaitara.movieapp.model.GetRecipesResponse$$serializer>: java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/String;Lkotlinx/serialization/internal/GeneratedSerializer;I)V in class Lkotlinx/serialization/internal/SerialClassDescImpl; or its super classes (declaration of 'kotlinx.serialization.internal.SerialClassDescImpl' appears in /data/app/com.janewaitara.movieapp-AkpKgLgxt6cvA-4-Zg5_7Q==/base.apk)```

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/RecipeApplication.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    companion object{
+
+        private lateinit var instance : RecipeApplication
+
+        lateinit var database: RecipeDatabase
+
+        fun getAppContext(): Context = instance.applicationContext
+
+        //exposing the remoteAPI to the rest of the app to the rest of the app(building and passing)
+        private val apiService by lazy { buildApiService() }
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/RecipeApplication.kt:20]: Nice job using the by lazy

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/db/RoomDatabase.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    @Delete
+    suspend fun deleteRecipe(recipe: Recipe)
+
+    @Update
+    fun updateRecipe(recipe: Recipe)
+
+}
+
+@Database(entities = [Recipe::class],version = 1, exportSchema = false)
+@TypeConverters(IngredientsConverter::class)
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/db/RoomDatabase.kt:43]: Nice work including the type converters for the list in the data.

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/model/IngredientsConverter.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.janewaitara.movieapp.model
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/model/IngredientsConverter.kt:1]: Nice job providing a type converter.

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/model/Recipe.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import android.os.Parcelable
+import androidx.room.ColumnInfo
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import kotlinx.serialization.Serializable
+
+@Serializable
+@Entity(tableName = "recipe_table")
+data class Recipe(
+    @PrimaryKey val id: Int,
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/model/Recipe.kt:13]: As Filip discussed, different model classes can be provided for different parts of the project to reduce the number of tags, but overall good work on this model class.

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/networking/RetrofitApiService.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.janewaitara.movieapp.networking
+
+import com.janewaitara.movieapp.model.GetRecipesResponse
+import retrofit2.http.GET
+import retrofit2.http.Query
+
+/**
+ * Represent the API calls you can make
+ * */
+interface RemoteApiService {
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/networking/RetrofitApiService.kt:10]: Nice job on the API Service class here.

File: Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/ui/login/LoginFragment.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import androidx.lifecycle.lifecycleScope
+import androidx.navigation.findNavController
+import com.janewaitara.movieapp.storage.RecipeSharedPrefs
+import com.janewaitara.movieapp.R
+import kotlinx.android.synthetic.main.fragment_login.*
+import kotlinx.coroutines.launch
+
+class LoginFragment : Fragment() {
+
+    private lateinit var loginPrefs : RecipeSharedPrefs
```
Comment (jenndotcodes) [inline Week 7/RecipeApp/app/src/main/java/com/janewaitara/movieapp/ui/login/LoginFragment.kt:17]: Watch out for the lateinit variables unless absolutely necessary.

**Review by jenndotcodes — APPROVED** [review]
Hi Jane, I signed up for an api key and put it in the gradle.properties file I created. It doesn't seem to be getting the data for me and I'm not encountering the error you listed, I did a search and thought something like this might be related to it https://github.com/Kotlin/kotlinx.serialization/issues/576. The app launches for me. The architechture looks good, watch out for the lateinit vars and avoid when possible. You did a very nice job identifying the list in your api's data and providing a type convertor for it and using coroutines. Let me know if I'm missing something, I'll award a 'good job', if the data loads and it's something on my end of things I'll update your score. Sorry I can't be more help on the error you mentioned, please don't hesitate to ask in the main bootcamp chat channel.

**Comment by janewaitara** [comment]
Hey Jenn,

I hope you are doing well and keeping safe.

Finally I have solved the bug after a week of research.The issue was being
caused by kotlin serialization as you had said cause of the list of
ingredients within the recipe.I have been forced to use moshi and all is
well now.

Thanks.

Jane Waitara

**Pull request merged**